server/aws_iid_attestor.go

```diff
@@ -0,0 +1,240 @@
+C1haGgSI/A1uZUKs/Zfnph0oEI0/hu1IIJ/SKBDtN5lvmZ/IzbOPIJWirlsllQIQ
+7zvWbGd9c9+Rm3p04oTvhup99la7kZqevJK0QRdD/6NpCKsqP/0=
+-----END CERTIFICATE-----`
+
+type IIDAttestorConfig struct {
+	TrustDomain string `hcl:"trust_domain"`
+}
+
+type IIDAttestorPlugin struct {
+	trustDomain string
+
+	awsCaCertPublicKey *rsa.PublicKey
+
+	mtx *sync.Mutex
+}
+
+func (p *IIDAttestorPlugin) spiffeID(awsAccountId, awsInstanceId string) *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, awsAccountId, awsInstanceId)
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host:   p.trustDomain,
+		Path:   spiffePath,
+	}
+	return id
+}
+
+func (p *IIDAttestorPlugin) Attest(req *nodeattestor.AttestRequest) (*nodeattestor.AttestResponse, error) {
+
+	var attestedData aia.IidAttestedData
+	err := json.Unmarshal(req.AttestedData.Data, &attestedData)
+	if err != nil {
+		err = aia.AttestationStepError("unmarshaling the attestation data", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	var doc aia.InstanceIdentityDocument
+	err = json.Unmarshal([]byte(attestedData.Document), &doc)
+	if err != nil {
+		err = aia.AttestationStepError("unmarshaling the IID", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	if req.AttestedBefore {
+		err = aia.AttestationStepError("validating the IID", fmt.Errorf("the IID has been used and is no longer valid"))
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	docHash := sha256.Sum256([]byte(attestedData.Document))
+	if err != nil {
+		err = aia.AttestationStepError("hashing the IID", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	sigBytes, err := base64.StdEncoding.DecodeString(attestedData.Signature)
+	if err != nil {
+		err = aia.AttestationStepError("base64 decoding the IID signature", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	err = rsa.VerifyPKCS1v15(p.awsCaCertPublicKey, crypto.SHA256, docHash[:], sigBytes)
+	if err != nil {
+		err = aia.AttestationStepError("verifying the cryptographic signature", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	awsSession := session.Must(session.NewSession())
+
+	ec2Client := ec2.New(awsSession, &aws.Config{
+		Region: &doc.Region,
+	})
+
+	query := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{&doc.InstanceId},
+	}
+
+	result, err := ec2Client.DescribeInstances(query)
+	if err != nil {
+		err = aia.AttestationStepError("querying AWS via describe-instances", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	instance := result.Reservations[0].Instances[0]
+
+	ifaceZeroDeviceIndex := *instance.NetworkInterfaces[0].Attachment.DeviceIndex
+
+	if ifaceZeroDeviceIndex != 0 {
+		innerErr := fmt.Errorf("DeviceIndex is %d", ifaceZeroDeviceIndex)
+		err = aia.AttestationStepError("verifying the EC2 instance's NetworkInterface[0].DeviceIndex is 0", innerErr)
```

Comment:
What does this error mean for a user? i.e., what as someone who isn't sure what this check is all about, I'm not sure what this message is telling me. Additionally, some code comments about these checks will help too